### PR TITLE
Vxlan6

### PR DESCRIPTION
**Pull request by outscale-mgo**
And here come the headache...

### CODE REVIEW

File: include/packetgraph/ip.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _PG_IP_H
+#define _PG_IP_H
+
+#include <rte_ip.h>
```
Comment (outscale-jju) [inline include/packetgraph/ip.h]: to remove

File: include/packetgraph/ip.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+union pg_ipv6_addr {
+	uint8_t word8[16];
+	uint16_t word16[8];
+	uint32_t word32[4];
+	uint64_t word64[2];
+} __attribute__((__packed__));
+
+
+
```
Comment (outscale-jju) [inline include/packetgraph/ip.h]: remove space

File: include/packetgraph/ip.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#define pg_ip_from_str(ip, src)						\
+	(inet_pton(							\
+		_Generic((ip), uint8_t *: AF_INET6, uint32_t : AF_INET), \
+		src,							\
+		_Generic((ip), uint8_t *: (ip), uint32_t : &(ip))	\
+		))
+
+
+
```
Comment (outscale-jju) [inline include/packetgraph/ip.h]: same here

File: include/packetgraph/ip.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#define PG_IPV6_HALF_FORMAT "%02x%02x:%02x%02x:%02x%02x:%02x%02x"
+
+#define PG_IPV6_FORMAT PG_IPV6_HALF_FORMAT":"PG_IPV6_HALF_FORMAT
+
+
```
Comment (outscale-jju) [inline include/packetgraph/ip.h]: same

File: include/packetgraph/ip.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+static inline const char *pg_ipv6_to_str(union pg_ipv6_addr *ip)
+{
+	static char  static_bufs[4][64];
```
Comment (outscale-jju) [inline include/packetgraph/ip.h]: remove ' ' + more buffers

File: src/utils/ip.h [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _PG_UTILS_IP_H
+#define _PG_UTILS_IP_H
+
+#include <rte_ip.h>
+#include <packetgraph/ip.h>
+
+
```
Comment (outscale-jju) [inline src/utils/ip.h]: remove space

File: src/utils/ip.h [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+{
+	return pg_is_same_ipv6_tab_tab(dest, src->word8);
+}
+
+static inline bool pg_is_same_ipv6_puni_puni(union pg_ipv6_addr *dest,
+					     union pg_ipv6_addr *src)
+{
+	return pg_is_same_ipv6_tab_tab(dest->word8, src->word8);
+}
+
```
Comment (outscale-jju) [inline src/utils/ip.h]: simplification to use only uint8_t* ?

File: src/utils/ip.h [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+	dst.addr_bytes[1] = 0x33;
+	dst.addr_bytes[2] = ip.word8[12];
+	dst.addr_bytes[3] = ip.word8[13];
+	dst.addr_bytes[4] = ip.word8[14];
+	dst.addr_bytes[5] = ip.word8[15];
+	return dst;
+}
+
+
+
```
Comment (outscale-jju) [inline src/utils/ip.h]: spaces

File: src/utils/ip.h [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+{
+	return ip->word8[0] == 0xff;
+}
+
+static inline bool is_multicast_ip_tab(uint8_t *ip)
+{
+	return ip[0] == 0xff;
+}
+
+
```
Comment (outscale-jju) [inline src/utils/ip.h]: space

File: src/utils/ip.h [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+ *
+ * @param	ip the ip to check, must be in network order (big indian)
+ * @return	1 if true, 0 if false
+ */
+static inline bool is_multicast_ipv4(uint32_t ip)
+{
+	return (((uint8_t *) &ip)[0] >= 224 && ((uint8_t *) &ip)[0] <= 239);
+}
+
+
```
Comment (outscale-jju) [inline src/utils/ip.h]: space

File: src/utils/errors.c [hunk 11]
```diff
@@ -61,8 +73,8 @@ void pg_error_free(struct pg_error *error)
 {
 	if (!error)
 		return;
-	g_free(error->context.file);
-	g_free(error->context.function);
+	g_free((char *)error->context.file);
+	g_free((char *)error->context.function);
```
Comment (outscale-jju) [inline src/utils/errors.c]: do we really need this cast ?

File: include/packetgraph/vtep.h [hunk 12]
```diff
@@ -54,10 +55,35 @@ struct ether_addr *pg_vtep_get_mac(struct pg_brick *brick);
...
-int pg_vtep_add_vni(struct pg_brick *brick,
-		    struct pg_brick *neighbor,
-		    uint32_t vni, uint32_t multicast_ip,
-		    struct pg_error **errp);
+int pg_vtep4_add_vni(struct pg_brick *brick,
+		     struct pg_brick *neighbor,
+		     uint32_t vni, uint32_t multicast_ip,
+		     struct pg_error **errp);
+
+
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: spaces

File: include/packetgraph/vtep.h [hunk 13]
```diff
@@ -54,10 +55,35 @@ struct ether_addr *pg_vtep_get_mac(struct pg_brick *brick);
...
+/**
+ * Add a VNI to the VTEP
+ *
+ * NOTE: Adding the same VNI twice is not authorized and will result in an
+ *       assertion
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: Is this still true ?
Maybe we have to fix this comment or fix this behavior by just setting errp

File: include/packetgraph/vtep.h [hunk 14]
```diff
@@ -54,10 +55,35 @@ struct ether_addr *pg_vtep_get_mac(struct pg_brick *brick);
...
+ * @param	vni the VNI to add, must be < 2^24
+ * @param	multicast_ip the multicast ip to associate to the VNI
+ * @param	errp an error pointer
+ */
+int pg_vtep6_add_vni(struct pg_brick *brick,
+		     struct pg_brick *neighbor,
+		     uint32_t vni, union pg_ipv6_addr multicast_ip,
+		     struct pg_error **errp);
+
+
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: space

File: include/packetgraph/vtep.h [hunk 15]
```diff
@@ -85,8 +111,36 @@ int pg_vtep_add_mac(struct pg_brick *brick, uint32_t vni,
...
+			      struct ether_addr mac, uint16_t udp_dst_port,
+			      int flag, struct pg_error **errp);
+
+/**
+ * Create a new vtep
+ *
+ * @name:	  brick name
+ * @max:	  maximum number of connections
+ * @output:	  side where packets are encapsuled in VXLAN
+ * @ip:		  vtep ip
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: ident fail

File: include/packetgraph/vtep.h [hunk 16]
```diff
@@ -85,8 +111,36 @@ int pg_vtep_add_mac(struct pg_brick *brick, uint32_t vni,
...
+			      struct ether_addr mac, uint16_t udp_dst_port,
+			      int flag, struct pg_error **errp);
+
+#define pg_vtep_new(name, max, output, ip, mac, udp_dst_port, flags, errp) \
+	_Generic((ip), uint32_t : pg_vtep4_new,				\
+		 int32_t : pg_vtep4_new,				\
+		 union pg_ipv6_addr : pg_vtep6_new)			\
+	(name, max, output, ip, mac, udp_dst_port, flags, errp)
+	
+
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: space

File: src/vtep-internal.c [hunk 17]
```diff
@@ -0,0 +1,913 @@
...
+ * by the Free Software Foundation.
+ *
+ * Butterfly is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butterfly.  If not, see <http://www.gnu.org/licenses/>.
+ *
```
Comment (outscale-jju) [inline src/vtep-internal.c]: Separate comments

```
/* LICENSE
 * ...
 */

/* comments
 * ...
 */
```

File: include/packetgraph/ip.h (hunk 5 above)
Comment (outscale-mgo) [inline include/packetgraph/ip.h]: '' ?

File: src/utils/ip.h (hunk 10 above)
Comment (outscale-mgo) [inline src/utils/ip.h]: https://www.youtube.com/watch?v=xeKMS62GrTI

File: configure_clang [hunk 18]
```diff
@@ -2,7 +2,6 @@
 export CC="clang"
 export AR="llvm-ar"
 export NM="llvm-nm"
-export LD="llvm-link"
```
Comment (outscale-jju) [inline configure_clang]: I don't quite understand this move :p

File: tests/style/test.sh [hunk 19]
```diff
@@ -74,7 +74,7 @@ else
     if [[ $major -lt 1 || $minor -lt 71 ]]; then
         echo "cppcheck version >= 1.71 required (currently have ${major}.${minor}), skip test"
     else
-        cppcheck -q  -f -I $PACKETGRAPH_ROOT/src -I $PACKETGRAPH_ROOT/include --error-exitcode=43 --enable=style --enable=performance --enable=portability --enable=information --enable=missingInclude --enable=warning $c_filelist
```
Comment (outscale-jju) [inline tests/style/test.sh]:
```
tests: remove style check from cppchak
-->
tests: remove style check from cppcheck
```

File: src/utils/mac-table.h [hunk 20]
```diff
@@ -163,7 +163,7 @@ static inline void pg_mac_table_ptr_unset(struct pg_mac_table *ma,
 	if (!pg_mac_table_is_set(*ma->ptrs[part1], part2))
 		return;
 	pg_mac_table_mask_unset(*ma->ptrs[part1], part2);
-	if (!ma->ptrs[part1]->mask) {
```
Comment (outscale-jju) [inline src/utils/mac-table.h]: would say
```
mac-table: fix pg_mac_table_ptr_unset 
-->
utils: fix pg_mac_table_ptr_unset 
```
as you already speak about "pg_mac_table_ptr_unset"

File: src/packets.c [hunk 21]
```diff
@@ -105,15 +105,15 @@ struct rte_mbuf **pg_packets_append_str(struct rte_mbuf **pkts,
 					const char *str)
 {
 	PG_PACKETS_OPS_STR(pkts, pkts_mask, str, append)
-	return pkts;
+		return pkts;
```
Comment (outscale-jju) [inline src/packets.c]: `return` and `PG_PACKETS_OPS_STR()` should be on the same level of indentation.
We can add a ";" at the end of `PG_PACKETS_OPS_STR()` so it do not fools the indent.

Comment (outscale-jju) [inline src/packets.c]: Also:
```
packets: reindent packets.c
-->
core: reindent packets.c
```

File: src/vtep.c [hunk 22]
```diff
@@ -135,7 +135,7 @@ struct vtep_state {
 	uint16_t udp_dst_port;		/* UDP destination port */
 	struct ether_addr mac;		/* MAC address of the VTEP */
 	struct vtep_port *ports;
-	rte_atomic16_t packet_id;	/* IP identification number */
+	uint16_t packet_id;	/* IP identification number */
```
Comment (outscale-jju) [inline src/vtep.c]: indentation ?

File: src/vtep.c [hunk 23]
```diff
@@ -344,8 +344,8 @@ static inline void ip_build(struct vtep_state *state, struct ipv4_hdr *ip_hdr,
 	ip_hdr->total_length = rte_cpu_to_be_16(datagram_len);
 
 	/* Set the packet id and increment it */
-	ip_hdr->packet_id =
-		rte_cpu_to_be_16(rte_atomic16_add_return(&state->packet_id, 1));
+	state->packet_id += 1;
+	ip_hdr->packet_id += rte_cpu_to_be_16(state->packet_id);
```
Comment (outscale-jju) [inline src/vtep.c]: I think there is still some atomic things here:
```
tests/vtep/tests.c:95:  rte_atomic16_t packet_id;       /* IP identification number */
```

File: src/vtep.c (hunk 22 above)
Comment (outscale-jju) [inline src/vtep.c]: also:
```
vtep: change satte->packet_id type from atomic to uint16_t 
-->
vtep: get ride of useless atomics
```

File: src/utils/ip.h [hunk 24]
```diff
@@ -0,0 +1,281 @@
+/* Copyright 2015 Outscale SAS
```
Comment (outscale-jju) [inline src/utils/ip.h]: 2017

File: src/utils/ip.h [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+		uint8_t version : 4;
+		uint8_t traffic_class : 8;
+		uint32_t flow_label : 20;
+	};
+	uint32_t vtc_flow;
+} __attribute__((__packed__));
+
+#define PG_IP_GENERIC_IPV6(callback)		\
+	uint8_t[16] : callback,			\
+		uint8_t * : callback
```
Comment (outscale-jju) [inline src/utils/ip.h]: ident ?

File: src/utils/ip.h [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+			   PG_IP_GENERIC_IPV6(pg_is_same_ipv6_tab_tab),	\
+			   union pg_ipv6_addr * : pg_is_same_ipv6_tab_puni) \
+		)(dest, src))
+
+static inline bool pg_is_same_ipv4(uint32_t a, uint32_t b)
+{
+	return a == b;
+}
+
+static inline bool pg_is_same_ipv6_tab_tab(uint8_t *dest, uint8_t *src)
```
Comment (outscale-jju) [inline src/utils/ip.h]: the convention is more like src, dst.

File: src/packets.h [hunk 27]
```diff
@@ -96,6 +96,12 @@ struct rte_mbuf **pg_packets_append_ipv4(struct rte_mbuf **pkts,
 					 uint16_t datagram_len,
 					 uint8_t proto);
 
+struct rte_mbuf **pg_packets_append_ipv6(struct rte_mbuf **pkts,
+					 uint64_t pkts_mask,
+					 uint8_t *src_ip, uint8_t *dst_ip,
+					 uint16_t datagram_len,
+					 uint8_t proto);
+
```
Comment (outscale-jju) [inline src/packets.h]: Ok,
just: you already have "packet" in the commit name so:
```
packets: add pg_packets_append_ipv6
-->
core: add pg_packets_append_ipv6 
```

File: include/packetgraph/vtep.h [hunk 28]
```diff
@@ -45,19 +48,61 @@ struct ether_addr *pg_vtep_get_mac(struct pg_brick *brick);
 /**
  * Add a VNI to the VTEP
  *
- * NOTE: Adding the same VNI twice is not authorized and will result in an
- *       assertion
```
Comment (outscale-jju) [inline include/packetgraph/vtep.h]: This warning is not true anymore ?

File: src/vtep6.c [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+
+
+static void multicast6_subscribe(struct vtep_state *state,
+				struct vtep_port *port,
+				union pg_ipv6_addr multicast_ip,
+				struct pg_error **errp)
+{
+	multicast6_internal(state, port, &multicast_ip, MLD_SUBSCRIBE,
+			    errp);
+}
```
Comment (outscale-jju) [inline src/vtep6.c]: Why not directly call multicast6_internal instead of multicast6_subscribe ?

File: src/vtep6.c [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+
+	.init		= vtep_init,
+	.destroy	= vtep_destroy,
+
+	.unlink		= pg_brick_generic_unlink,
+
+	.unlink_notify  = vtep_unlink_notify,
+};
+
+pg_brick_register(vtep, &vtep_ops);
```
Comment (outscale-jju) [inline src/vtep6.c]: Don't forget the [checklist](https://github.com/outscale/packetgraph/blob/master/doc/contrib.md#checklist), I think you need to complete `README.md` ;)

File: tests/vtep/tests.c [hunk 31]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+}
+
+static void test_vtep6_simple(void)
+{
+	struct ether_addr mac_dest = {{0xb0, 0xb1, 0xb2,
+				      0xb3, 0xb4, 0xb5} };
+	struct ether_addr mac_src = {{0xc0, 0xc1, 0xc2,
+				      0xc3, 0xc4, 0xc5} };
+	struct ether_addr multicast_mac1, multicast_mac2;
+	/*we're usin the same cfg for the collect and the hub*/
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: add spaces

File: tests/vtep/tests.c [hunk 32]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	struct pg_error *error = NULL;
+	struct ether_hdr *tmp;
+	union pg_ipv6_addr ip_src = {.word64 = {0,1}};
+	union pg_ipv6_addr ip_dst = {.word64 = {0,2}}; 
+	union pg_ipv6_addr ip_vni1 = {.word16 = {0xff,0,0,0,0,0,0,0x1}};
+	union pg_ipv6_addr ip_vni2 = {.word16 = {0xff,0,0,0,0,0,0,0x2}};
+	int flag = 0;
+	uint16_t i;
+
+	/*For testing purpose this vtep ip is 1*/
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: add spaces + 'For -> for`

File: tests/vtep/tests.c [hunk 33]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	vtep_west = pg_vtep6_new("encapsulatron", 500, PG_EAST_SIDE,
```
Comment (outscale-jju) [inline tests/vtep/tests.c]:
```
encapsulatron
-->
encapsulation
```

File: tests/vtep/tests.c [hunk 34]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+				 ip_src.word8, mac_src, PG_VTEP_DST_PORT,
+				 flag, &error);
+	g_assert(!error);
+	g_assert(vtep_west);
+
+	/*For testing purpose this vtep ip is 2*/
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: add spaces + `For -> for`

File: tests/vtep/tests.c [hunk 35]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	g_assert(!error);
+	pg_brick_link(collect_west2, vtep_west, &error);
+	g_assert(!error);
+	pg_brick_link(hub, collect_hub, &error);
+	g_assert(!error);
+	pg_brick_link(vtep_east, collect_east2, &error);
+	g_assert(!error);
+
+
+	graph = pg_graph_new("big dad", collect_west1, &error);
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: What ?

File: tests/vtep/tests.c [hunk 36]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	g_assert(!error);
+
+	pg_scan_ether_addr(&multicast_mac1, "01:00:5e:00:00:05");
+	pg_scan_ether_addr(&multicast_mac2, "01:00:5e:00:00:06");
+	/* Configure VNI */
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: C->c

File: tests/vtep/tests.c [hunk 37]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	pg_error_make_ctx(1);
+	check_collector(collect_hub, pg_brick_west_burst_get, 1);
+	pg_vtep6_add_vni(vtep_east, collect_east2, 1,
+			 ip_vni2.word8, &error);
+	g_assert(!error);
+	pg_error_make_ctx(1);
+	check_collector(collect_hub, pg_brick_west_burst_get, 1);
+
+	pg_vtep6_add_vni(vtep_west, collect_west1, 0,
+		      ip_vni1.word8, &error);
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: ident ?

File: tests/vtep/tests.c [hunk 38]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	g_assert(!error);
+	pg_error_make_ctx(1);
+	check_collector(collect_hub, pg_brick_west_burst_get, 1);
+
+	pg_vtep6_add_vni(vtep_west, collect_west2, 1,
+		      ip_vni2.word8, &error);
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: ident ?

File: tests/vtep/tests.c [hunk 39]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	mac_src.addr_bytes[5] = 0xf5;
+
+	mac_dest.addr_bytes[0] = 0xe0;
+	mac_dest.addr_bytes[1] = 0xe1;
+	mac_dest.addr_bytes[2] = 0xe2;
+	mac_dest.addr_bytes[3] = 0xe3;
+	mac_dest.addr_bytes[4] = 0xe4;
+	mac_dest.addr_bytes[5] = 0xe5;
+
+	/* Alloc packets */
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: A -> a

File: tests/vtep/tests.c [hunk 40]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+
+	/* check the packets have been recive by collect_west1 */
+	pg_error_make_ctx(1);
+	check_collector(collect_west1, pg_brick_east_burst_get, NB_PKTS);
+
+	/* check no packets have been recive by collect_west2 */
+	pg_error_make_ctx(1);
+	check_collector(collect_west2, pg_brick_east_burst_get, 0);
+
+
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: space

File: tests/vtep/tests.c [hunk 41]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+		g_assert(is_same_ether_addr(&tmp->s_addr, &mac_dest));
+		g_assert(is_same_ether_addr(&tmp->d_addr, &mac_src));
+		g_assert(!hdr->udp.dgram_cksum);
+	}
+
+	pg_brick_burst_to_east(vtep_west, 0, pkts,
+			    pg_mask_firsts(NB_PKTS), &error);
+	g_assert(!error);
+
+
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: space

File: tests/vtep/tests.c [hunk 42]
```diff
@@ -126,6 +154,236 @@ static struct ether_addr multicast_get_dst_addr(uint32_t ip)
...
+	/* check the packets have been recive by collect_west1 */
+	pg_error_make_ctx(1);
+	result_pkts = check_collector(collect_hub, pg_brick_west_burst_get,
+				      NB_PKTS);
+
+	pg_error_make_ctx(1);
+	check_collector(collect_east1, pg_brick_west_burst_get,
+			NB_PKTS);
+
+
```
Comment (outscale-jju) [inline tests/vtep/tests.c]: space

File: configure_clang (hunk 18 above)
Comment (outscale-mgo) [inline configure_clang]: well, llvm-link seem to be buggy on my machine, but clang fail to link examples with ld, so I can't use clang with lld on my machine, but clang travis need lld, so my solution is to put lld in travis file, and let the user choose his compiler

File: tests/vtep/tests.c (hunk 33 above)
Comment (outscale-mgo) [inline tests/vtep/tests.c]: Well, encapsulatron is the brick name, "encapsulation" is not really a name, encaplulator might be a better name ?

File: tests/vtep/tests.c (hunk 35 above)
Comment (outscale-mgo) [inline tests/vtep/tests.c]: well the graph can be see as the bricks fathers right ?

**Pull request merged**